python/cuml/numba_utils.py

```diff
@@ -50,6 +70,30 @@ def kernel(_col_major, _row_major):
             _row_major[tid, col_idx] = _col_major[tid, col_idx]
             _col_offset += 1
 
-    kernel[bpg, tpb](col_major, row_major)
+    @cuda.jit
+    def shared_kernel(input, output):
+
+        tile = cuda.shared.array(shape=tile_shape, dtype=dev_dtype)
+
+        tx = cuda.threadIdx.x
+        ty = cuda.threadIdx.y
+        bx = cuda.blockIdx.x * cuda.blockDim.x
+        by = cuda.blockIdx.y * cuda.blockDim.y
+        y = by + tx
+        x = bx + ty
+
+        if by + ty < input.shape[0] and bx + tx < input.shape[1]:
+            tile[ty, tx] = input[by + ty, bx + tx]
+        cuda.syncthreads()
+        if y < output.shape[0] and x < output.shape[1]:
+            output[y, x] = tile[tx, ty]
+
+    # check if we cannot call the shared memory kernel
+    # block limits: 2**31-1 for x, 65535 for y dim of blocks
+    if blocks[0] > 2147483647 or blocks[1] > 65535:
```

Comment:
The shared memory route is so much faster. Since transposition is a bijective function, can't this be batched for the case of higher block sizes? For instance, if each kernel is working on pieces of the larger array, we should be able to adjust the output positions in the kernel for each piece to account for that without having to use any additional locks for atomics, right? Maybe something for 0.8?

Reply:
Excellent point! Something to look into for 0.8, thought maybe first I can have an estimate for which dimensions the non shared kernel gets run (and therefore a proxy of what percentage of cases go through thw slower route) to see if the priority of the task :)